adapter/src/main/java/com/twilio/kudu/sql/NestedJoinFactory.java

```diff
@@ -0,0 +1,140 @@
+/* Copyright 2021 Twilio, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.twilio.kudu.sql;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.apache.calcite.linq4j.AbstractEnumerable;
+import org.apache.calcite.linq4j.Enumerable;
+import org.apache.calcite.linq4j.Enumerator;
+import org.apache.calcite.linq4j.Linq4j;
+import org.apache.calcite.linq4j.function.EqualityComparer;
+import org.apache.calcite.linq4j.function.Function1;
+import org.apache.calcite.linq4j.function.Predicate2;
+import org.apache.kudu.client.AsyncKuduScanner;
+
+/**
+ * {@code NestedJoinFactory} is an implementation of {@link Function1} that
+ * returns an {@link Enumerable} for the Right hand side of a nested join.
+ * <p>
+ * This implementation uses a cache to hold objects from the scan to prevent
+ * going creating {@link AsyncKuduScanner} for repeated rows
+ */
+public final class NestedJoinFactory implements Function1<List<Object>, Enumerable<Object>> {
+
+  private final KuduRowCache rowCache;
+
+  private final Predicate2<Object, Object> joinCondition;
+
+  private final List<TranslationPredicate> rowTranslators;
+
+  private final KuduEnumerable rootEnumerable;
+
+  public NestedJoinFactory(final Predicate2<Object, Object> joinCondition, final EqualityComparer<Object> compareUtil,
+      final List<TranslationPredicate> rowTranslators, final KuduEnumerable rootEnumerable) {
+    this.joinCondition = joinCondition;
+    this.rowCache = new KuduRowCache(compareUtil, 1000);
+    this.rowTranslators = rowTranslators;
+    this.rootEnumerable = rootEnumerable;
+  }
+
+  @Override
+  public Enumerable<Object> apply(final List<Object> batchFromLeftTable) {
```

Comment:
Looking at the code it looks like the size of `batchFromLeftTable` is 100. So for each batch of 100 rows from the left size we would first check the right size cache of 1000 and if the rows are not found we would then create scans for right side?

Reply:
> Looking at the code it looks like the size of `batchFromLeftTable` is 100. So for each batch of 100 rows from the left size we would first check the right size cache of 1000 and if the rows are not found we would then create scans for right side?

We would create scans for the `misses`. So if there are two in the cache and 98 not in the cache, it constructs scans for the 98 that are missing.

Reply:
Should we still have the processForInList optimization so that it issues 1 scan for the rows that aren't in the cache instead of a scan per missing row?